File: LeetCode75_Artur/Heap and Priority Queue/215_Kth_Largest_Element_in_an_Array.py
"""
https://leetcode.com/problems/kth-largest-element-in-an-array/?envType=study-plan-v2&envId=leetcode-75
"""

from typing import Optional, List
import heapq

class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        heap = nums[:k]
        heapq.heapify(heap)
        
        for i in range(k, len(nums)):
            if nums[i] > heap[0]:
                heapq.heappop(heap)
                heapq.heappush(heap, nums[i])
                
        return heap[0]
    
        
if (__name__ == '__main__'):
    solution = Solution()
    
    test1 = [3, 2, 1, 5, 6, 4]
    test2 = [3, 2, 3, 1, 2, 4, 5, 5, 6]
    
    print(solution.findKthLargest(test1, 2))
    print(solution.findKthLargest(test1, 1))
    print(solution.findKthLargest(test1, 6))
    print(solution.findKthLargest(test1, 4))
    
    print(solution.findKthLargest(test2, 2))
    print(solution.findKthLargest(test2, 1))
    print(solution.findKthLargest(test2, 9))
    print(solution.findKthLargest(test2, 4))
    
"""
Будуємо піраміду з k елементів. Основна ідея полягає в тому,
щоб у цій піраміді містити k найбільших елементів, а оскільки
нам потрібно знайти k-ий найбільший елемент, то ця піраміда буде
min-heap, тому на її вершині міститиметься необхідний елемент

Після того, як додали перших k елементів в піраміду запускаємо 
heapify, що отримати мінімальний елемент на вершині.

Починаємо ітеруватись з k елементу масиву і якщо цей елемент є більшим
за вершину піраміди, то вершину видаляємо з неї, а цей елемент пушимо
в піраміду.

У кінці в піраміді міститиметься k найбільших елементів, тому повертаємо
нульовий, який є найменшим серед них.

Часова складність: O(nlogn)
Додавання елементу в піраміду займає logn, а в найгіршому випадку ми цю 
операцію викличемо n разів, тома загальна складність алгоритму O(nlogn).

Також цікавий спосіб за O(n)

class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        min_value = min(nums)
        max_value = max(nums)
        count = [0] * (max_value - min_value + 1)
        for num in nums:
            count[num - min_value] += 1

        remain = k
        for i in range(len(count) - 1, - 1, -1):
            remain -= count[i]
            if remain <= 0:
                return i + min_value
        return -1
        
Спершу шукаємо наймеший та найбільший елемент, створюємо масив, який
буде відповідати за кількість елементів певного значення в даному 
масиві, це масив містить max - min елементів.

Підраховуємо всі елементи, а дальше йдемо з кінця та віднімаємо кількість
входжень всіх елементів від k до тих пір, поки це значення не буде меншим 
або рівним k.

Яка проблема цієї реалзіації?
Якщо буде дуже велика різниця між мінімальним та максимальним значенням, то
масив для підрахунків входжень буде надто великим, а також якщо елементів у
заданому масиву буде мало, то це ще й займе дуже багато часу на обхід цього 
масиву, що по часу може програвати іншим реалізаціям.
Також виникнуть проблеми, якщо числа будуть не цілими.

Тому для підрахунку краще використовувати мапу.
"""

File: LeetCode75_Artur/Heap and Priority Queue/215_Kth_Largest_Element_in_an_Array.cpp
/*
 https://leetcode.com/problems/kth-largest-element-in-an-array/?envType=study-plan-v2&envId=leetcode-75
*/

#include <iostream>
#include <vector>
#include <algorithm>


using namespace std;

class Solution
{
public:

    int findKthLargest(vector<int>& nums, int k)
    {
        sort(nums.begin(), nums.end());
        return nums[nums.size() - k];
    }
};


int main()
{
    Solution solution;

    vector<int> test1 = {3, 2, 1, 5, 6, 4};
    vector<int> test2 = {3, 2, 3, 1, 2, 4, 5, 5, 6};

    cout << solution.findKthLargest(test1, 2) << endl;
    cout << solution.findKthLargest(test1, 1) << endl;
    cout << solution.findKthLargest(test1, 3) << endl;
    cout << solution.findKthLargest(test1, 5) << endl;

    cout << solution.findKthLargest(test2, 2) << endl;
    cout << solution.findKthLargest(test2, 1) << endl;
    cout << solution.findKthLargest(test2, 3) << endl;
    cout << solution.findKthLargest(test2, 5) << endl;

}

/*
 Main idea:
 Загалом для цього завдання є три алгоритми його розв'язку:
 - посортувати і вернути k-найбільший елемент — тут він
   реалізований вище;
 - через minheap — його детальний опис і реалізація буде в
   пайтонівському варіанті
 - QuickSelect — нижче написав свою реалізацію через quickselect,
   але з ним 40 тест не проходить по часу, навіть з рандомним
   півотом, оскільки там майже всі елементи повторюються і попадається
   випадок зі складність O(n^2).

 Також одна сторінка з рішень цієї задачі із поясненнями плюсів та мінусів
 цих підходів:

 https://leetcode.com/problems/kth-largest-element-in-an-array/solutions/3906260/100-3-approaches-video-heap-quickselect-sorting/?envType=study-plan-v2&envId=leetcode-75

 class Solution
 {
 public:
    int partition(vector<int>& nums, int left, int right)
    {
        srand(time(0));
        int randPivot = rand() % (right - left + 1) + left;
        swap(nums[randPivot], nums[right]);

        int pivot = nums[right];
        int i = left;

        for(int j = left; j < right; ++j)
            if( nums[j] <= pivot )
                swap(nums[i++], nums[j]);

        swap(nums[i], nums[right]);
        return i;
    }

    int quickSelect(vector<int>& nums, int left, int right, int k)
    {
        if( left == right )
            return nums[right];

        int q = partition(nums, left, right);
        int i = q - left + 1;
        if (i == k)
            return nums[q];
        else if( k < i )
            return quickSelect(nums, left, q - 1, k);
        return quickSelect(nums, q + 1, right, k - i);
    }

    int findKthLargest(vector<int>& nums, int k)
    {
        return quickSelect(nums, 0, (int) nums.size() - 1, (int) nums.size() - k + 1);
    }
 };
*/